cpp_exercise/src/lib/ipcshmlib.cc

```diff
@@ -0,0 +1,247 @@
+    
+
+}
+SharedMemorySender::~SharedMemorySender(){
+
+
+     
+    sem_close(_sem_sender);
+    sem_close(_sem_receiver);
+    munmap(_ptr,_shm_size);
+    shm_unlink(_shm_name.c_str());
+}
+
+
+SharedMemoryReceiver::SharedMemoryReceiver(const std::string filename): _file_name(filename){
+    
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"****************************IPCReceiver: SharedMemory PROTOCOL**************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+        std::cout<<"**************************************************************"<<std::endl;
+
+ 
+
+   
+
+    
+     sem_unlink(_semaphoresender_name.c_str());
+     sem_unlink(_semaphorereceiver_name.c_str());
+     shm_unlink(_shm_name.c_str());
+
+
+     _sem_sender = sem_open(_semaphoresender_name.c_str(),O_CREAT ,0660,0);
+
+     if(_sem_sender == SEM_FAILED){
+
+        std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCReceiver ERROR:Cannot Open Semaphore Sender.");
+     }
+     
+     _sem_receiver = sem_open(_semaphorereceiver_name.c_str(),O_CREAT , 0660,1);
+   
+  
+     if(_sem_receiver == SEM_FAILED){
+
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCReceiver ERROR:Cannot Open Semaphore Receiver.");
+     }
+
+ 
+    std::cout<<"          Waiting for the Sender..."<<std::endl;
+
+   _shm_fd = shm_open(_shm_name.c_str(), O_CREAT | O_RDONLY, 0666);
+     if(_shm_fd < 0){
+         std::cout<<strerror(errno)<<std::endl;
+         throw IPCException("IPCReceiver ERROR:Cannot Open Shared Memory.");
+     }
+
+  
+   _ptr =(_shm_data_struct *) mmap(0, _shm_size, PROT_READ, MAP_SHARED, _shm_fd, 0);
+      
+    if(_ptr == (struct _shm_data_struct *)-1){
+        std::cout<<strerror(errno)<<std::endl;
+        throw IPCException("IPCReceiver ERROR: Cannot map the Shared Memory.");
+    }
+    
+
+
+    FileHandler fr(_file_name);
+    fr.createFile();
+    
+    std::cout<<"          Starting to read the data: "<<std::endl; 
+
+    while(true){
+
+      
+               
+            if (clock_gettime(CLOCK_REALTIME, &_ts) == -1)
+            {
+                std::cout<<strerror(errno)<<std::endl;
+                 throw IPCException("IPCReceiver ERROR: Cannot get the CLOCK TIME.");
+            }
+
+            _ts.tv_sec += 15;
+    
+            int s =  sem_timedwait(_sem_sender,&_ts);
+
+             if (s == -1)
+            {
+            if (errno == ETIMEDOUT){
+
+            break;
+            }
+            }
+
+   
+            _shm_data_struct tempstruct = *_ptr;
+                
+            if(tempstruct.datasize == 0){
+
+                break;
+            }
+      
+            std::vector<unsigned char> tempvec(_buffer_size);
+            tempvec.insert(tempvec.begin(),std::begin(tempstruct.data),std::end(tempstruct.data));
```

Comment:
In this case I'd see some sense in intermediate copy as write to file can be performance bottleneck, but to benefit from it writeFile shouldn't be guarded by semaphore

Reply:
I think i fixed it by moving the writeFile out of the critical section